lib/matplotlib/collections.py

```diff
@@ -2016,6 +2015,55 @@ def set_paths(self):
             self._meshWidth, self._meshHeight, self._coordinates)
         self.stale = True
 
+    def set_array(self, A):
+        width, height = self._meshWidth, self._meshHeight
+        misshapen_data = False
+        faulty_data = False
+        
+        if A is not None:
+            if self._shading == 'flat':
+                if A.ndim == 1:
+                    if A.size not in (height*width, (height + 1)*(width + 1)):
+                        faulty_data = True
+                elif A.shape not in ((height, width), (height + 1, width + 1)):
+                    if A.size in (height * width, (height + 1) * (width + 1)):
+                        misshapen_data = True
+                    else:
+                        faulty_data = True
+
+            elif self._shading == 'nearest':
+                if A.ndim == 1:
+                    if A.size != width * height:
+                        faulty_data = True
+                elif A.shape != (height, width):
+                    if A.size == height * width:
+                        misshapen_data = True
+                    else:
+                        faulty_data = True
+
+            elif self._shading == 'gouraud':
+                if A.ndim == 1:
+                    if A.size != (width + 1) * (height + 1):
+                        faulty_data = True
+                elif A.shape != (height + 1, width + 1):
+                    if A.size == (height + 1) * (width + 1):
+                        misshapen_data = True
+                    else:
+                        faulty_data = True
```

Comment:
I appologize if I mislead you before.  This can all be much simpler.

QuadMesh only takes `flat`: C[M, N], X[N+1], Y[M+1] or `gouraud`: C[M, N], X[N], Y[M].  

This is not tested, but I think covers all the cases....
```suggestion
        if A is not None:
            if self._shading == 'flat':
                if A.size != height * width:
                    faulty_data = True
                if A.shape != (height, width):
                    misshapen_data = True

            elif self._shading == 'gouraud':
                if A.size != (height + 1) * (width + 1):
                    faulty_data = True
                if A.shape != (height + 1, width + 1):
                    misshapen_data = True
```

Please make sure to add a test for gouraud shading.

Reply:
Thanks for the explanation, I'll add the tests accordingly :)

Reply:
If its not the right total size its faulty and we error.  If its not the right shape its misshapen and we warn...  Though actually you may need to put the raise before the warn so we don't warn and then raise

Reply:
Yes, but 1-D arrays should be accepted without an explicit warning right? I thought we only raise a warning if the size of the array matches but the dimensions don't **and** they are of more than 1 Dimensions.

Reply:
I guess that was left unresolved above.  I'd say we should deprecate 1d as well.  @timhoffm ?

Reply:
Yes, since we are deprecating 1d input in __init__ we should deprecate that here as well. Plea see use `_api.warn_deprecated()` for that.

Reply:
But 1d arrays are passed to set_array by default right because `rave()` is called under the hood?
Deprecating this would require a change over here as well?
https://github.com/matplotlib/matplotlib/blob/78e3b5f8f12500510fd06567a12bb70f42608f9f/lib/matplotlib/axes/_axes.py#L6122

Reply:
I think that is right.  We discussed QuadMesh on the weekly call, and there is no need for the ravel behaviour that any of the lead devs know of.  We might as well fix it all now.  This may break some downstream libraries who patterned artists on `pcolormesh` and also included the ravel, but it is just confusing for a fundamentally 2-D object to always take a ravel.

Reply:
Okay, apologies for my asking but in removing that line causes a lot of other tests to fail.

Reply:
This is more complex than anticipated.

Changing the shape of the internal representation of "value-like" quantities such as `C` is non-trivial. If we do so, that should apply to all such quantities (see all parameters that accept lists in `Collection`). That includes shape-checking in __init__/setters, docstring updates, changed return values for getters. Also we have to check if `Collection` can handle multiple dimensions out of the box - as it's written/documented it's not clear if that will work.

For now, I'm ok, if this PR only checks the input to accept (M*N, 2) and (M, N, 2). All other inputs with (M*N*2) are deprecated and reshaped to (M*N, 2). The internal representation stays (M*N, 2).

Further investigation into the API is welcome, but I anticipate that it will be a significant effort.

Reply:
Ah, but as of now, are we deprecating the 1-D input? I'm still a bit confused as to if that has to be changed as well.

Reply:
Sorry for being confusing. I've been partially mixing this with the coordinates discussion.

With this PR, we should
- accept a 2D array-like of shape (M, N)
- accept a 1D array-like of length M*N
- deprecate supporting any differently shaped array-like with a total of M*N elements.
- raise on any other array-like.

where the mesh consists of M rows and N columns of quadrilaterals.

Reply:
Yeah this makes sense, thanks for explaining it.
Sorry for the confusion this PR caused :sweat_smile: